File: src/services/notes.ts
import { invoke } from "@tauri-apps/api/core";
import {
  noteListItemSchema,
  noteRevisionSchema,
  noteSchema,
  type CreateNoteInput,
  type Note,
  type NoteListItem,
  type NoteRevision,
  type UpdateNoteInput,
} from "../types/note";

export const noteService = {
  async create(input: CreateNoteInput = {}): Promise<Note> {
    const value = await invoke("create_note", { input });
    return noteSchema.parse(value);
  },

  async get(id: string): Promise<Note> {
    const value = await invoke("get_note", { id });
    return noteSchema.parse(value);
  },

  async list(includeArchived = false): Promise<NoteListItem[]> {
    const value = await invoke("list_notes", { includeArchived });
    return noteListItemSchema.array().parse(value);
  },

  async search(query: string): Promise<NoteListItem[]> {
    const value = await invoke("search_notes", { query });
    return noteListItemSchema.array().parse(value);
  },

  async listDeleted(): Promise<NoteListItem[]> {
    const value = await invoke("list_deleted_notes");
    return noteListItemSchema.array().parse(value);
  },

  async update(input: UpdateNoteInput): Promise<Note> {
    const value = await invoke("update_note", { input });
    return noteSchema.parse(value);
  },

  async archive(id: string, archived: boolean): Promise<Note> {
    const value = await invoke("archive_note", { id, archived });
    return noteSchema.parse(value);
  },

  async remove(id: string): Promise<void> {
    await invoke("delete_note", { id });
  },

  async restore(id: string): Promise<Note> {
    const value = await invoke("restore_note", { id });
    return noteSchema.parse(value);
  },

  async permanentlyRemove(id: string): Promise<void> {
    await invoke("permanently_delete_note", { id });
  },

  async revisions(noteId: string): Promise<NoteRevision[]> {
    const value = await invoke("list_note_revisions", { noteId });
    return noteRevisionSchema.array().parse(value);
  },
};

File: src-tauri/src/commands/notes.rs
use crate::{
    db::{
        models::{CreateNoteInput, Note, NoteListItem, NoteRevision, UpdateNoteInput},
        repository::NoteRepository,
    },
    error::AppError,
};
use tauri::State;

#[tauri::command]
pub async fn create_note(
    repository: State<'_, NoteRepository>,
    input: CreateNoteInput,
) -> Result<Note, AppError> {
    repository.create(input).await.map_err(Into::into)
}

#[tauri::command]
pub async fn search_notes(
    repository: State<'_, NoteRepository>,
    query: String,
) -> Result<Vec<NoteListItem>, AppError> {
    repository.search(&query).await.map_err(Into::into)
}

#[tauri::command]
pub async fn list_deleted_notes(
    repository: State<'_, NoteRepository>,
) -> Result<Vec<NoteListItem>, AppError> {
    repository.list_deleted().await.map_err(Into::into)
}

#[tauri::command]
pub async fn get_note(repository: State<'_, NoteRepository>, id: String) -> Result<Note, AppError> {
    repository.get(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn list_notes(
    repository: State<'_, NoteRepository>,
    include_archived: bool,
) -> Result<Vec<NoteListItem>, AppError> {
    repository.list(include_archived).await.map_err(Into::into)
}

#[tauri::command]
pub async fn update_note(
    repository: State<'_, NoteRepository>,
    input: UpdateNoteInput,
) -> Result<Note, AppError> {
    repository.update(input).await.map_err(Into::into)
}

#[tauri::command]
pub async fn archive_note(
    repository: State<'_, NoteRepository>,
    id: String,
    archived: bool,
) -> Result<Note, AppError> {
    repository.archive(&id, archived).await.map_err(Into::into)
}

#[tauri::command]
pub async fn delete_note(
    repository: State<'_, NoteRepository>,
    id: String,
) -> Result<(), AppError> {
    repository.soft_delete(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn restore_note(
    repository: State<'_, NoteRepository>,
    id: String,
) -> Result<Note, AppError> {
    repository.restore_deleted(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn permanently_delete_note(
    repository: State<'_, NoteRepository>,
    id: String,
) -> Result<(), AppError> {
    repository.permanently_delete(&id).await.map_err(Into::into)
}

#[tauri::command]
pub async fn list_note_revisions(
    repository: State<'_, NoteRepository>,
    note_id: String,
) -> Result<Vec<NoteRevision>, AppError> {
    repository.revisions(&note_id).await.map_err(Into::into)
}
